### ISSUE TITLE
"threads.c" error when running binaries or docker image for amd64 devel version

### ISSUE THREAD

**Issue by laurenceisla** [PostgREST/postgrest#3569]
### Environment

* PostgreSQL version: PostgreSQL 15.6 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 12.3.0, 64-bit
* PostgREST version: v12.1 (devel)
* PostgREST docker version: devel
* Operating system: NixOS 23.11

### Description of issue

Executing the binary or docker release [for the `devel` branch ](https://github.com/PostgREST/postgrest/releases/tag/devel), returns an error when trying to connect to the database.

Local binary:

```
PGRST_DB_URI=postgres://postgres@localhost:5432/postgres \
PGRST_DB_ANON_ROLE=postgres \
PGRST_DB_LOG_LEVEL=info \
postgrest-run
```
```console
03/Jun/2024:13:19:01 -0500: Starting PostgREST 12.1...
03/Jun/2024:13:19:01 -0500: Listening on port 3000
03/Jun/2024:13:19:01 -0500: Attempting to connect to the database...
Assertion failed: *(&once->n) != 4 (threads.c: k5_once: 564)
fish: Job 1, 'PGRST_DB_URI=postgres://postgre…' terminated by signal PGRST_DB_ANON_ROLE=postgres \ (PGRST_DB_LOG_LEVEL=info \)
fish: Job ./postgrest, 'SIGABRT' terminated by signal Abort ()
```

Docker:

```
sudo docker run --net=host --rm --name test-db-conn -p 3000:3000 \
            -e PGRST_DB_URI="postgres://postgres@localhost:5432/postgres" \
            -e PGRST_DB_ANON_ROLE="postgres" \
            -e PGRST_LOG_LEVEL="info" \
            postgrest/postgrest:devel
```
```console
03/Jun/2024:18:21:03 +0000: Starting PostgREST 12.1...
03/Jun/2024:18:21:03 +0000: Listening on port 3000
03/Jun/2024:18:21:03 +0000: Attempting to connect to the database...
Assertion failed: *(&once->n) != 4 (threads.c: k5_once: 564)
```

Running the `main` branch from a Nix environment does not return this error:

```bash
PGRST_DB_URI=postgres://postgres@localhost:5432/postgres \
PGRST_DB_ANON_ROLE=postgres \
PGRST_DB_LOG_LEVEL=info \
postgrest-run
```
```console
03/Jun/2024:13:18:45 -0500: Starting PostgREST 12.1 (b6c6f2b)...
03/Jun/2024:13:18:45 -0500: Attempting to connect to the database...
03/Jun/2024:13:18:45 -0500: Listening on port 3000
03/Jun/2024:13:18:45 -0500: Successfully connected to PostgreSQL 15.6 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 12.3.0, 64-bit
...
```

**Comment by steve-chavez** [comment]
> Local binary:

```
PGRST_DB_URI=postgres://postgres@localhost:5432/postgres \
PGRST_DB_ANON_ROLE=postgres \
PGRST_DB_LOG_LEVEL=info \
postgrest-run
```

There seems to be a mistake above, `postgrest-run` doesn't work but I can reproduce with the binary. Like so:

```
PGRST_DB_URI=postgres://postgres@localhost:5432/postgres \
PGRST_DB_ANON_ROLE=postgres \
PGRST_DB_LOG_LEVEL=info \
./postgrest
```

**Comment by steve-chavez** [comment]
> Can reproduce with the downloaded binary and a local PG (non nix tooled)

Same for me, it doesn't happen when using the nix `postgrest-with-postgresql-*` pg, only when using my distro postgres.

**Comment by steve-chavez** [comment]
Error comes from kerberos lib: https://github.com/krb5/krb5//blob/5f0023d5f05e95021a7caa1193f76f86871222ce/src/util/support/threads.c#L558-L565

Maybe the error starts from https://github.com/PostgREST/postgrest/pull/3169?

**Comment by steve-chavez** [comment]
Ah damn, it likely is the exception throwing and catching I did https://github.com/PostgREST/postgrest/blob/a46bea16e239d120804d7d125f1f737e4bfb3b8d/src/PostgREST/Listener.hs#L45-L53

Somehow it must be affecting some state on the C lib. I will try a different approach.

Why this doesn't happen on the local nix postgres though? Because it connects through unix sockets?

**Comment by laurenceisla** [comment]
> Why this doesn't happen on the local nix postgres though? Because it connects through unix sockets?

I don't think so. Using `postgrest-run` in the nix environment but connecting to my local database over TCP works OK (it's my last example).

**Comment by steve-chavez** [comment]
So it wasn't the exception catching/throwing, I just removed that and it fails the same.

Another change on https://github.com/PostgREST/postgrest/commit/3cf565614d6529b4f691699b254fd3a680dfcda0 is that now the listener and the connection worker start in parallel.

Before the listener waited for the connection worker, here https://github.com/PostgREST/postgrest/commit/3cf565614d6529b4f691699b254fd3a680dfcda0#diff-f6074642dc32a94e9e1997b21b0df631b2d3f098e0a81673edc22b89a86f49edL534-L538. But that was only done to reuse the exponential backoff retrying from the connection worker on the listener, I didn't know it would have this effect. I guess the kerberos lib doesn't work with parallel connections?

**Comment by steve-chavez** [comment]
Oh wait.. I have a fix right now that brings back waiting between listener and connection worker..

**Comment by steve-chavez** [comment]
Yeah, I think it's worth adding it back. If anything it reduces wastefully retrying both listener and the connection worker when postgres is completely down. Doing it on https://github.com/PostgREST/postgrest/pull/3573.

>  maybe when two pool connections start up at the same time or something like that.

Right, that could still be trouble.
